posts/alpine-stack-size.md

```diff
@@ -0,0 +1,155 @@
+  /usr/local/bundle/gems/rspec-core-3.7.1/lib/rspec/core/reporter.rb:171:in `finish'
+  /usr/local/bundle/gems/rspec-core-3.7.1/lib/rspec/core/reporter.rb:81:in `report'
+  /usr/local/bundle/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:112:in `run_specs'
+  /usr/local/bundle/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:87:in `run'
+  /usr/local/bundle/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:71:in `run'
+  /usr/local/bundle/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:45:in `invoke'
+  /usr/local/bundle/gems/rspec-core-3.7.1/exe/rspec:4:in `<top (required)>'
+  /usr/local/bundle/bin/rspec:29:in `load'
+  /usr/local/bundle/bin/rspec:29:in `<top (required)>'
+```
+
+This was extremely strange since we did not notice this error locally. 
+
+## Tracing the error
+
+### Increasing the stack size
+
+At first we thought there was an environment configuration difference between the Ruby in our CI pipeline and our local machines.
+
+So we attempted to increase the stack-size with the CI environment via the `RUBY_THREAD_VM_STACK_SIZE` environment variable.
+
+This had no effect.
+
+We thought this was very odd, since we still could not reproduce locally.
+
+### Decreasing the stack size
+
+We then decided to attempt to reduce the stack size on our local machines (again with the `RUBY_THREAD_VM_STACK_SIZE` environment variable).
+
+What we discovered was that we needed to have stack sizes <500KB in order to reproduce the issue.
+
+This got us no closer to a fix, but did tell us that somehow the stack size was being restricted in our CI environment.
+
+### Running with Alpine locally
+
+Gaining some parity at this point made a lot of sense. While we didn't expect there to be a runtime difference within Ruby between local and Alpine, it was definitely worth ruling it out.
+
+Immediately we noticed the error locally.
+
+```
+SystemStackError: stack level too deep
+```
+
+## What was wrong?
+
+Doing some google searching we ended up at musl-libc's FAQs, specifically the ["functional differences"](https://wiki.musl-libc.org/functional-differences-from-glibc.html#Thread-stack-size) i.e. incompatibilities between glibc. 
+
+What we learned is that musl provides a default stack size of 80k, which is a significant difference to the stacksize provided by glibc (generally 2MB-10MB).
+
+Moreover, it requires programs to explicitly ask for more (or less) stack via `pthread_attr_setstacksize`, which MRI Ruby does not appear to do.
+
+There is an open Ruby bug here: https://bugs.ruby-lang.org/issues/14387
+
+## How did we solve it?
+
+We submitted a [patch upstream to RSpec](https://github.com/rspec/rspec-support/pull/343) to change the implementation from a recursive algorithm to a loop-based algorithm that does not use the stack.
+
+Ultimately this is a temporary fix, as any recursive logic could be a source of errors in production. This should be solved at the Ruby language level.
+
+## Design problems
+
+This is a liskov substition principle violation.
```

Comment:
I would tend to capitalise Liskov (Wiki does on its page - see below).

Also, typo on `substitution` - here and l140.

![image](https://user-images.githubusercontent.com/429326/38800824-27e88502-4160-11e8-86cc-ff3e080e0349.png)